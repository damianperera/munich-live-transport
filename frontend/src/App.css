.app {
  text-align: center;
}

.blinking {
  color: green; /* Set text color to green */
  animation: blink 2s infinite; /* Blinking animation */
}

@keyframes blink {
  0% {
    opacity: 1; /* Fully visible for the first second */
  }
  50% {
    opacity: 1; /* Fully visible for the second half of the cycle */
  }
  100% {
    opacity: 0; /* Invisible for the second second */
  }
}

.departures-container {
  position: relative;
  width: calc(100vw - 20px); /* Fill the horizontal space, subtracting 40px for left and right margin */
  height: calc(100vh - 20px); /* Fill the vertical space, subtracting 40px for top and bottom margin */
  margin: 10px; /* Adjust margin as needed */
}

.departures {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-collapse: collapse;
  border: none; /* Thick border */
  margin: 0; /* Remove margin */
  padding: 0; /* Remove padding */
}

.departures td {
  border: 2px solid darkgray; /* Thick border */
  padding: 3px; /* Reduce padding */
  white-space: nowrap; /* Prevent text wrapping */
  overflow: hidden; /* Hide overflowing content */
  text-overflow: ellipsis; /* Display ellipsis for truncated text */
  font-size: calc(3vw + 10px); /* Set font size based on screen width */
}

.departures td:first-child {
  max-width: 2vw !important;
}

.departures td:nth-last-child(2) {
  font-size: calc(3vw + 10px);
  padding-left: 2%; /* Add padding to the left */
  max-width: 10vw !important;
}

.departures td:last-child {
  max-width: 6vw !important; /* 2/5ths of the screen width, subtracting border width */
  position: relative;
}

.center {
  text-align: center;
}

.left {
  text-align: left;
}

.right {
  text-align: right;
  padding-right: 1% !important;
}

/* Responsive styles */
@media screen and (max-width: 600px) {
  .departures td {
    max-width: 200px; /* Limit width of cells */
  }
}

.departureNested {
  max-width: 25% !important;
  width: 25% !important;
  min-width: 25% !important;
  text-align: center ;
  font-size: calc(1vw + 3px) !important;
  position: absolute;
  margin-left: calc(1vw - 13px);
  border: none !important;
  padding-top: 1%;
  padding-left: 3% !important;
  padding-right: 3% !important;
  padding-bottom: 1% !important;
  line-height: inherit;
  height: 1.5vw;
}

.departureSev {
  background-color: purple !important;
  color: white !important;
  margin-top: calc(2vw + 2px);
  font-weight: bold;
}

.departureCancelled {
  background-color: red !important;
  color: white;
  font-weight: bold;
  margin-top: calc(0.5vw - 7px);
  z-index: 999;
}

.departureDelayed {
  background-color: #ff8503;
  color: white;
  font-weight: bold;
  margin-top: calc(0.5vw - 7px);
}

.loading-container {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(255, 255, 255, 0.5); /* Semi-transparent background */
  z-index: 9998; /* Ensure it's on top */
}

.loading-text {
  margin: 0;
  font-size: 3vw;
  color: black; /* Text color */
  margin-left: 10px; /* Adjust spacing between text and spinner */
}

.loading-spinner {
  border: 4px solid darkgray; /* Light grey border */
  border-top: 4px solid darkblue; /* Blue border */
  border-radius: 50%;
  width: 40px;
  height: 40px;
  animation: spin 2s linear infinite; /* Animation */
  margin-right: 10px; /* Adjust spacing between spinner and text */
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}

.error-container {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(255, 255, 255); /* Semi-transparent background */
  z-index: 9997; /* Ensure it's on top */
}

.error-text {
  margin: 0;
  font-size: 3vw;
  color: black; /* Text color */
  margin-left: 10px; /* Adjust spacing between text and spinner */
}

.error-text small {
  font-size: 2vw;
  text-align: center;
}

.red {
  color: red;
}

.settings-container {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: center;
  align-items: center;
  z-index: 9999;
}

.settings {
  background-color: white;
  padding: 1%;
  border-radius: 8px;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5);
  width: 90vw; /* Set width relative to viewport width */
  max-width: 400px; /* Set maximum width for larger screens */
}

.settings h2 {
  margin: 1%;
  font-size: 1.5vw;
}

.settings h2 small {
  font-weight: normal;
}

.settings-content {
  display: flex;
  flex-direction: column;
}

.settings-input {
  font-size: 1vw;
  padding: 1.5%;
  text-align: left;
}

.settings-buttons {
  display: flex;
  justify-content: flex-end; /* Align buttons to the right */
  flex-wrap: wrap;
}

.settings-buttons button {
  margin: 1.5%;
  padding-left: 3%;
  padding-right: 3%;
  padding-top: 2%;
  padding-bottom: 2%;
  border-radius: 4px;
  cursor: pointer;
  color: black;
  flex-grow: 1;
  background-color: white;
}

.settings-buttons button:last-child {
  border-color: #ff8503;
}

.settings-buttons button:nth-last-child(2) {
  border-color: green;
}

.settings-buttons button:nth-last-child(3) {
  border-color: red;
}

.settings-buttons button:nth-last-child(4) {
  border-color: blueviolet;
}

.settings-buttons button:first-child {
  border-color: gray;
}

.license {
  margin-top: 2%;
  font-size: 1vw;
  border-top: 1px solid lightgray;
  padding-top: 2%;
}

.license a {
  text-decoration: none;
}

.license pre {
  margin: 0 !important;
  display: inline !important;
  background-color: rgba(211, 211, 211, 0.5);
  border-radius: 3px;
  padding: 2px;
  cursor: pointer;
}
