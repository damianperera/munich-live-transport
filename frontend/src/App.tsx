import React, { useState, useEffect } from 'react';
import './App.css';

// type DisruptionStationProps = {
//   id: string;
//   name: string;
// }

// type DisruptionLineProps = {
//   id: string;
//   name: string;
//   typeOfTransport: string;
//   stations: DisruptionStationProps[];
//   direction: string;
// }

// type DisruptionProps = {
//   id: string;
//   type: string;
//   title: string;
//   text: string;
//   lines: DisruptionLineProps[];
// }

type DepartureStationProps = {
  type: string;
  latitude: string;
  longitude: string;
  place: string;
  name: string;
  globalId: string;
  divaId: string;
  transportTypes: [];
  surroundingPlanLink: string;
  aliases: string;
  tariffZones: string;
}

type DepartureProps = {
  plannedDepartureTime: string;
  realtime: boolean;
  delayInMinutes: number;
  realtimeDepartureTime: number;
  transportType: string;
  label: string;
  divaId: string;
  network: string;
  trainType: string;
  destination: string;
  cancelled: boolean;
  sev: boolean;
  platform: number;
  platformChanged: boolean;
  messages: any[];
  bannerHash: string;
  occupancy: string;
  stopPointGlobalId: string;
}

type TransformedDepartureProps = {
  label: string;
  destinations: {
    destination: string;
    departures: DepartureProps[];
  }[];
}

function App() {
  const DEFAULT_STATION = 'Forstenrieder Allee'
  const TYPE_STATION = 'STATION'
  const TYPE_UBAHN = 'UBAHN'
  // const [departureStations, setDepartureStation] = useState<DepartureStationProps>()
  const [departures, setDepartures] = useState<TransformedDepartureProps[]>([])
  // const [disruptions, setDisruptions] = useState<DisruptionProps[]>([])

  useEffect(() => {
    const getDepartureStation = async () => {
      let data = await fetch(`https://www.mvg.de/api/fib/v2/location?query=${encodeURI(DEFAULT_STATION)}`, {
        method: "GET"
      });
      let jsonData = await data.json();

      const pds = jsonData.find((e: DepartureStationProps) => e.type === TYPE_STATION)
      data = await fetch(`https://www.mvg.de/api/fib/v2/departure?globalId=${pds?.globalId}&limit=20&offsetInMinutes=0`, {
        method: "GET"
      });
      const departures: DepartureProps[] = await data.json();
      const sortedDepartures: TransformedDepartureProps[] = transformDepartures(departures)

      // setDepartureStation(pds)
      setDepartures(sortedDepartures);
    }

    // const getDisruptions = async () => {
    //   const data = await fetch("https://www.mvg.de/api/ems/tickers", {
    //     method: "GET"
    //   });
    //   const jsonData = await data.json();
    //   setDisruptions(jsonData);
    // };

    const runEveryMinute = () => {
      getDepartureStation();
      // getDisruptions();

      setTimeout(runEveryMinute, 60000);
    };

    const timeoutId = setTimeout(runEveryMinute, 0)
    return () => clearTimeout(timeoutId);
  }, []);

  const calculateRemainingTime = (epochTime: number) => {
    const currentTime = new Date().getTime();
    const remainingMilliseconds = epochTime - currentTime;

    if (remainingMilliseconds < 60000) {
      return (<strong className="blinking">now</strong>);
    }

    const remainingMinutes = Math.floor(remainingMilliseconds / 60000);
    const hours = Math.floor(remainingMinutes / 60);
    const minutes = remainingMinutes % 60;

    let remainingTimeString = '';
    if (hours > 0) {
      remainingTimeString += `${hours} ${hours === 1 ? 'hr' : 'hrs'}`;
      if (minutes > 0) {
        remainingTimeString += `${minutes} ${minutes === 1 ? 'min' : 'mins'}`;
      }
    } else {
      remainingTimeString += `${minutes} ${minutes === 1 ? 'min' : 'mins'}`;
    }

    return (<span>{remainingTimeString}</span>);
  };

  function transformDepartures(departures: DepartureProps[]): TransformedDepartureProps[] {
    const transformedArray = departures.reduce((acc: TransformedDepartureProps[], departure: DepartureProps) => {
      const { label, destination } = departure;
      const labelIndex = acc.findIndex(item => item.label === label);
      if (labelIndex === -1) {
        acc.push({ label, destinations: [{ destination, departures: [departure] }] });
      } else {
        const destIndex = acc[labelIndex].destinations.findIndex(item => item.destination === destination);
        if (destIndex === -1) {
          acc[labelIndex].destinations.push({ destination, departures: [departure] });
        } else {
          acc[labelIndex].destinations[destIndex].departures.push(departure);
        }
      }
      return acc;
    }, []);

    // Sort the transformed array so that 'UBAHN' departures come first
    transformedArray.sort((a, b) => {
      const aHasUBahn = a.destinations.some(dest => dest.departures.some(dep => dep.transportType === TYPE_UBAHN));
      const bHasUBahn = b.destinations.some(dest => dest.departures.some(dep => dep.transportType === TYPE_UBAHN));
      if (aHasUBahn && !bHasUBahn) {
        return -1;
      } else if (!aHasUBahn && bHasUBahn) {
        return 1;
      } else {
        return 0;
      }
    });

    return transformedArray.map(item => ({
      label: item.label,
      destinations: item.destinations.map(dest => ({
        destination: dest.destination,
        departures: dest.departures.sort((a, b) => a.realtimeDepartureTime - b.realtimeDepartureTime).slice(0, 2)
      }))
    }));
  }

  return (
    <div className="App">
      <header className="App-header">
        <div className='departures-container'>
          <table className='departures'>
            <tbody>
              {departures.map((item, index) => {
                return (
                  <React.Fragment key={index}>
                    {item.destinations.map((dest, destIndex) => {
                      return dest.departures.map((departure, depIndex) => {
                        return (
                          <tr key={`${index}-${destIndex}-${depIndex}`}>
                            {depIndex === 0 && (
                              <React.Fragment>
                                <td rowSpan={dest.departures.length}>
                                  {depIndex === 0 ? item.label : null}
                                </td>
                                <td className='left' rowSpan={dest.departures.length}>
                                  {dest.destination}
                                </td>
                              </React.Fragment>
                            )}
                            <td className='left'>{calculateRemainingTime(departure.realtimeDepartureTime)}</td>
                          </tr>
                        );
                      });
                    })}
                  </React.Fragment>
                );
              })}
            </tbody>
          </table>
        </div>
        {/* {disruptions.map((disruption, idx) => {
          return (
            <div key={idx}>
              <div><strong>{disruption.type}</strong></div>
              <div>{disruption.title}</div>
              <div>{disruption.text}</div>
            </div>
          );
        })} */}
      </header>
    </div>
  );
}

export default App;
