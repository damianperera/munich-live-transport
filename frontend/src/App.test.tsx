import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import App from './App'

test('app skeleton renders correctly', () => {
	const { container } = render(<BrowserRouter><App /></BrowserRouter>)
	expect(container.firstChild).toMatchSnapshot()
})