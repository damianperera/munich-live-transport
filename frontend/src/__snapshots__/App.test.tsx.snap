// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="App"
>
  <header
    class="App-header"
  >
    <div
      class="loading-container"
    >
      <div
        class="loading-spinner"
      />
      <p
        class="loading-text"
      >
        Loading Departures
      </p>
    </div>
    <div
      class="departures-container"
    >
      <table
        class="departures"
      >
        <tbody />
      </table>
    </div>
  </header>
</div>
`;
